java/arcs/android/host/ArcHostHelper.kt

```diff
@@ -0,0 +1,210 @@
+        // Ignore other actions
+        if (intent?.action?.startsWith(ARC_HOST_INTENT) != true) return
+
+        // Ignore Intent when it doesn't target our Service
+        if (intent.component?.equals(ComponentName(service, service::class.java)) != true) return
+
+        when (intent.getSerializableExtra(OPERATION) as Operation?) {
+            Operation.START_ARC -> runWithResult(intent) {
+                arcHost.startArc(
+                    intent.getParcelableExtra<ParcelablePlanPartition?>(
+                        OPERATION_ARG
+                    )?.actual!!
+                )
+            }
+            Operation.STOP_ARC -> runWithResult(intent) {
+                arcHost.stopArc(
+                    intent.getParcelableExtra<ParcelablePlanPartition?>(
+                        OPERATION_ARG
+                    )?.actual!!
+                )
+            }
+            Operation.REGISTER_PARTICLE -> runWithResult(intent) {
+                arcHost.registerParticle(
+                    intent.getParcelableExtra<ParcelableParticleIdentifier?>(
+                        OPERATION_ARG
+                    )?.actual!!
+                )
+            }
+            Operation.UNREGISTER_PARTICLE -> runWithResult(intent) {
+                arcHost.unregisterParticle(
+                    intent.getParcelableExtra<ParcelableParticleIdentifier?>(
+                        OPERATION_ARG
+                    )?.actual!!
+                )
+            }
+            Operation.GET_REGISTERED_PARTICLES -> runWithResult(intent) {
+                arcHost.registeredParticles()
+            }
+            else -> {
+                throw IllegalArgumentException("Operation $intent not implemented.")
+            }
+        }
+    }
+
+    /**
+     * Execute an [ArcHost] method in the block, encode return into a [Bundle] and send via
+     * [ResultReceiver]
+     */
+    private suspend fun runWithResult(intent: Intent, block: suspend () -> Any?) {
+        val bundle = Bundle()
+        val result = block()
+
+        when (result) {
+            is List<*> -> {
+                val arrayList = ArrayList<ParcelableParticleIdentifier>()
+                result.forEach {
+                    it as ParticleIdentifier
+                    arrayList += it.toParcelable()
+                }
+                bundle.putParcelableArrayList(OPERATION_RESULT, arrayList)
+            }
+            is Parcelable -> bundle.putParcelable(OPERATION_RESULT, result)
+            is Serializable -> bundle.putSerializable(OPERATION_RESULT, result)
+            else -> {
+                // Unit
+            }
+        }
+        // This triggers a suspended coroutine to resume with the value.
+        intent.getParcelableExtra<ResultReceiver>(OPERATION_RECEIVER)?.send(0, bundle)
+    }
+
+    enum class Operation {
+        START_ARC, STOP_ARC, REGISTER_PARTICLE, UNREGISTER_PARTICLE,
+        GET_REGISTERED_PARTICLES
+    }
```

Comment:
Consider placing on separate lines, and giving some KDoc, or restricting the visibility if it's not intended that ArcHost implementers need to know/access these.